File: src/main/java/AoC2022_06.java
import static com.github.pareronia.aoc.Range.range;

import java.util.List;

import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2022_06 extends AoCBase {
    
    private final String buffer;
    
    private AoC2022_06(final List<String> input, final boolean debug) {
        super(debug);
        this.buffer = input.get(0);
    }
    
    public static final AoC2022_06 create(final List<String> input) {
        return new AoC2022_06(input, false);
    }

    public static final AoC2022_06 createDebug(final List<String> input) {
        return new AoC2022_06(input, true);
    }
    
    private int solve(final int size) {
        for (final int i : range(size, buffer.length(), 1)) {
            if (range(i - size, i, 1).stream().map(buffer::charAt).distinct().count() == size) {
                return i;
            }
        };
        throw new IllegalStateException("Unsolvable");
    }
    
    @Override
    public Integer solvePart1() {
        return solve(4);
    }

    @Override
    public Integer solvePart2() {
        return solve(14);
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2022_06.createDebug(TEST1).solvePart1() == 7;
        assert AoC2022_06.createDebug(TEST2).solvePart1() == 5;
        assert AoC2022_06.createDebug(TEST3).solvePart1() == 6;
        assert AoC2022_06.createDebug(TEST4).solvePart1() == 10;
        assert AoC2022_06.createDebug(TEST5).solvePart1() == 11;
        assert AoC2022_06.createDebug(TEST1).solvePart2() == 19;
        assert AoC2022_06.createDebug(TEST2).solvePart2() == 23;
        assert AoC2022_06.createDebug(TEST3).solvePart2() == 23;
        assert AoC2022_06.createDebug(TEST4).solvePart2() == 29;
        assert AoC2022_06.createDebug(TEST5).solvePart2() == 26;

        final Puzzle puzzle = Aocd.puzzle(2022, 6);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_06.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_06.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST1 = splitLines("mjqjpqmgbljsphdztnvjfqwrcgsmlb");
    private static final List<String> TEST2 = splitLines("bvwbjplbgvbhsrlpgdmjqwftvncz");
    private static final List<String> TEST3 = splitLines("nppdvjthqldpwncqszvftbrmjlhg");
    private static final List<String> TEST4 = splitLines("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg");
    private static final List<String> TEST5 = splitLines("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw");
}

File: src/main/cpp/2022/06/AoC2022_06.cpp
#include <assert.h>

#include <string>
#include <unordered_set>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;
using namespace aoc;

int solve(const string& buffer, const size_t size) {
    for (int i : aoc::Range::range(size, buffer.size())) {
        const string& sub = buffer.substr(i - size, size);
        const unordered_set<char> s(sub.begin(), sub.end());
        if (s.size() == size) {
            return i;
        }
    };
    throw "unsolvable";
}

int part1(const vector<string>& input) { return solve(input[0], 4); }

int part2(const vector<string>& input) { return solve(input[0], 14); }

// clang-format off
const vector<string> TEST1 = {"mjqjpqmgbljsphdztnvjfqwrcgsmlb"};
const vector<string> TEST2 = {"bvwbjplbgvbhsrlpgdmjqwftvncz"};
const vector<string> TEST3 = {"nppdvjthqldpwncqszvftbrmjlhg"};
const vector<string> TEST4 = {"nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg"};
const vector<string> TEST5 = {"zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw"};
// clang-format on

void samples() {
    assert(part1(TEST1) == 7);
    assert(part1(TEST2) == 5);
    assert(part1(TEST3) == 6);
    assert(part1(TEST4) == 10);
    assert(part1(TEST5) == 11);
    assert(part2(TEST1) == 19);
    assert(part2(TEST2) == 23);
    assert(part2(TEST3) == 23);
    assert(part2(TEST4) == 29);
    assert(part2(TEST5) == 26);
}

MAIN(2022, 6)
